### PR TITLE
Node local dynamic filter

### PR DESCRIPTION
**Pull request by sopel39**

### CODE REVIEW

**Comment by sopel39** [comment]
Supersedes https://github.com/prestosql/presto/pull/931

**Comment by sopel39** [comment]
Benchmark results: https://github.com/prestosql/presto/pull/931#issuecomment-522909579

**Comment by sopel39** [comment]
Follow up work:
1. We need to add tests for `io.prestosql.sql.planner.LocalDynamicFilter`
2. We need to investigate: 
```
It seems that when dynamic filtering is enabled on Hive, the inner-join changes PARTITIONED to REPLICATED, seems to be related to the different partition keys used for ss_sold_date_sk.
```
3. We need to investigate regressions in benchmarks (q8 for example)

File: presto-main/src/main/java/io/prestosql/split/PageSourceManager.java [hunk 1]
```diff
@@ -47,18 +49,34 @@ public void removeConnectorPageSourceProvider(CatalogName catalogName)
...
-        return getPageSourceProvider(catalogName).createPageSource(
-                table.getTransaction(),
-                session.toConnectorSession(catalogName),
-                split.getConnectorSplit(),
-                table.getConnectorHandle(),
-                columns);
+        ConnectorPageSourceProvider provider = getPageSourceProvider(catalogName);
+        TupleDomain<ColumnHandle> constraint = TupleDomain.all();
+        if (dynamicFilter != null) {
+            constraint = dynamicFilter.get(); // should not block
```
Comment (lxynov) [inline presto-main/src/main/java/io/prestosql/split/PageSourceManager.java]: @rzeyde-varada @sopel39 
Thank you for your node local dynamic filtering work. It's really helpful.

> `// should not block`

Though I didn't have the context before, I've spent some time trying to understand the code. From my (coarse) understanding, the `ScanFilterAndProjectOperator` doesn't wait for the dynamic filter to be collected. So it is possible that the `constraint` we get here is `null` even if the dynamic filter is being actively collected on the hash build side. Is this understanding correct?

Comment (rzeyde-varada) [reply presto-main/src/main/java/io/prestosql/split/PageSourceManager.java]: You're welcome :)

> So it is possible that the `constraint` we get here is `null` even if the dynamic filter is being actively collected on the hash build side. 

Yes, this is indeed possible.

Comment (lxynov) [reply presto-main/src/main/java/io/prestosql/split/PageSourceManager.java]: Thanks for the clarification!

**Pull request merged**